[com_fields] Normalise the request com_fields data

laoneo opened the pull request:
Pull Request for Issues like #19735 and #17801 and for PRs #18188, #17837, #18207, #19742, #19753, #19850.

### Summary of Changes
As #19753 revealed that the request data should be normalised only for com_fields and we do not want to have extension specific code in the libraries. This pr adds a new event which sends the data for normalisation in the controller. Like that the fields plugin can then modify the data.

_Remark:_
I was unsure about the plugin name, any feedback is welcome.

### Testing Instructions
- Create custom user fields.
- Edit a user and fill any data in custom fields.
- Add the following code to the file /administrator/components/com_content/content.php after line 16: 
`JFactory::getUser($USER_ID)->save();`.
Replace the value $USER_ID with the user you have edited.

### Expected result
The user custom field data is still saved on the user.

### Actual result
The user custom fields data on the user are deleted.

laoneo committed:
Normalise the request com_fields data

laoneo committed:
CS

laoneo committed:
PHP 5.3 compat

ggppdk commented:
value '0' is quite common,
the checks empty($value) and empty($value[0]) should be changed

laoneo commented:
I'm also not sure if strlen is the solution as asked here https://github.com/joomla/joomla-cms/pull/19753#discussion_r172101916.

ggppdk commented:
> I'm also not sure if strlen is the solution as asked here

well, $value can not be an object, it is either
null, boolean, string, or array

null, boolean, string are all of them 'valid' inputs for strlen

so the only check that needs to be made is that $value and $value[0] are not an array

consider "empty" if
`is_array($value) ? !count($value) : !strlen($value)`
the above should replace `empty($value)`

`is_array($value[0]) ? !count($value[0]) : !strlen($value[0])`
the above should replace `empty($value[0])`

Quy commented on libraries/src/MVC/Controller/FormController.php:
```diff
@@ -694,6 +694,14 @@ public function save($key = null, $urlVar = null)
 			return false;
 		}
 
+		// Send an object which can be modified trough the plugin event
```
spelling `through`

Quy commented on plugins/system/fields/fields.php:
```diff
@@ -78,17 +111,35 @@ public function onContentAfterSave($context, $item, $isNew, $data = array())
+				$value = $field->rawvalue;
+			}
+
+			// If no value set (empty) remove value fom database
```
spelling `from`

Ruud68 commented on plugins/system/fields/fields.php:
```diff
@@ -30,6 +31,38 @@ class PlgSystemFields extends JPlugin
+			}
+
+			// Set a default value for the field
+			$data->{$field->fieldname} = false;
```
This is incorrect as the fields are not directly in the data, but in the variable 'com_fields':
this is the correct code:
```
	public function onContentNormaliseRequestData($context, $data, Form $form)
	{
		if (!FieldsHelper::extract($context, $data))
		{
			return true;
		}

		if (!isset($data->com_fields))
		{
			$data->com_fields = array();
		}

		// Loop over all fields
		foreach ($form->getGroup('com_fields') as $field)
		{
			// Make sure the data object has an entry
			if (isset($data->com_fields[$field->fieldname]))
			{
				continue;
			}

			// Set a default value for the field
			$data->com_fields[$field->fieldname] = false;
		}
	}
```

Ruud68 replied:
just did a PR for you :)

Ruud68 commented on libraries/src/MVC/Controller/FormController.php:
```diff
@@ -694,6 +694,14 @@ public function save($key = null, $urlVar = null)
 
+		// Send an object which can be modified trough the plugin event
+		$objData = (object) $data;
+		$app->triggerEvent(
```
tested this with both com_content and com_users: in the back-end it works, BUT when editing the user's profile via the front-end this does not get triggered? You can try this by login in on the front-end end then edit your own profile.

ggppdk replied:
The event does not get triggered because some controllers like the controller of user's profile
`components\com_users\controllers\profile.php`

do not call `parent::save()`

so these controllers that do not call parent method,
will need to be updated to include the code that triggers the event

i do not know if there are more controllers like this,
i think in frontend there are no more such controllers
with a quick search i did not find more in frontend,
i found in backend e.g. found template controllers but these do not make use of custom fields

e.g. in frontend
`D:\xampp\htdocs\devel\joomla-cms\components\com_content\controllers\article.php`
does not need to be updated with the plugin triggering code because it calls parent's save() method

Ruud68 replied:
Okay, thanks @ggppdk just did a PR against @laoneo 's repo for the ./components/com_users/controllers/profile.php

There is also a save function in register.php (when a user registers) Not sure if that also needs to be adjusted?

Ruud68 replied:
and I found the submit function in ./components/com_contact/controller/contact.php

those are the three (com_content, com_user, com_contact) components that use com_fields.

for register.php I am not sure because when you register, you don't have any values in com_fields so you can't loose them when saving.
For contact.php the same I guess, but I never use that component so not sure here

Ruud68 committed:
Fields in com_fields array (#9)

Fields should be set in com_fields array and not direcly in $data

laoneo committed:
Spelling

Ruud68 committed:
Also normalise request data on front-end user profile save (#10)

* Also normalise request data on front-end user profile save

* correct context and option

Ruud68 commented:
I have tested this item :white_check_mark: successfully on 5db8f95bdf141e7f4715f55c3d4408b912feb641<br /><br />Tested on both com_content and com_user (back-end and front-end)
1. saving with empty checkbox removes checkbox previous value from table = OK
2. removing value from field removes value from table = OK
3. setting field to only display on administrator and the save article or user on front-end retains value for the field in the table = OK
4. changing user groupmembership via API retains field values for this user = OK

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19884">issues.joomla.org/tracker/joomla-cms/19884</a>.</sub>

Ruud68 commented:
pinging @Denitz , @joomdonation , @ggppdk on this PR.

I have been trying to test for all your (related) issues and I think that they are fixed with this PR.
Can you test as well so we can get 'the show on the road' :)

Also calling on @tassosm and @regularlabs as I know they develop custom fields and (could) have an interest in also getting this fixed :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19884">issues.joomla.org/tracker/joomla-cms/19884</a>.</sub>

Quy commented:
@laoneo As @ggppdk has [commented](https://github.com/joomla/joomla-cms/pull/19884#issuecomment-372054240), the value of `0` will be cleared and not saved which I have confirmed.

laoneo committed:
Handle 0 properly in empty check

laoneo committed:
Simplify

laoneo commented:
Added the proposal from @ggppdk. Can you guys please test again?

Ruud68 commented:
Hi @laoneo , it fails :(
I have a text field and a checkbox field. It only saves one of them: when changing the checkbox, then the text field is deleted and vv. ~~reverting $needsDelete in https://github.com/joomla/joomla-cms/pull/19884/commits/9c68d5b9c7e9388dfda0f98de5911edff80b2a3c fixes it ($needsDelete = empty($value[0]);)~~

EDIT: fixed this via new PR against your repo

Ruud68 commented:
Hi @laoneo just did two PR's against your repo.
with these PR's it is possible to save value 0 as actual value AND it corrects the incorrect behaviour on single character text values (strlen = 1)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19884">issues.joomla.org/tracker/joomla-cms/19884</a>.</sub>

Ruud68 committed:
allowing value 0 to be saved (#11)

when setting a value of 0 in a text field the function empty will return true > setting the value to null

Quy commented:
Now clicking the `Save` button two times will clear entered values.

Ruud68 commented:
@Quy could you try this: https://github.com/Digital-Peak/joomla-cms/pull/12
That should fix it.

Quy commented:
@Ruud68 Fixed. Thank you!

ggppdk commented:
Ok please let me explain things 
the empty() check was in 2 places
now both are patched

but

@Ruud68 

It was:
> $needsUpdate = !empty($value[0]);

It was patched to be
> $needsUpdate = is_array($value[0]) ? !count($value[0]) : !strlen($value[0]);

The negation was of `!empty($value[0])` 
- was lost

> Now clicking the Save button two times will clear entered values.

that is why it is not working !

the complex expression that you suggest in your PR fixes it but it is not needed be like that,
you are just readding the negation back with a more complex expression

so just re-add the negation:
```php
$needsUpdate = ! (is_array($value[0]) ? !count($value[0]) : !strlen($value[0]));
//or better
$needsUpdate = is_array($value[0]) ? count($value[0]) : strlen($value[0]));
```

Ruud68 commented:
>the complex expression that you suggest in your PR fixes it but it is not needed be like that,
you are just readding the negation back with a more complex expression

Agree and @laoneo already simplified that. the code I suggested was a quick-mockup from me trouble shooting where things went wrong :)

With this last PR https://github.com/Digital-Peak/joomla-cms/pull/12 it works :)
with your code:
```
$needsUpdate = is_array($value[0]) ? count($value[0]) : strlen($value[0]);
```

When this PR is committed I will be re-doing all my test cases (which are already a lot :))

Ruud68 committed:
correct needsUpdate when strlen (or count) = 1 which incorrectly equa… (#12)

* correct needsUpdate when strlen (or count) = 1 which incorrectly equaled to 'true'

* Update field.php

* Update field.php

laoneo commented:
Pr's from @Ruud68 got merged.

Ruud68 commented:
I have tested this item :white_check_mark: successfully on 2756bb8ba6307f20f5769e7ff582379a037330b6<br /><br />Tested on both com_content and com_user (back-end and front-end)

1. saving with empty checkbox removes checkbox previous value from table = OK
2. removing value from field removes value from table = OK
3. setting field to only display on administrator and the save article or user on front-end retains value for the field in the table = OK
4. changing user groupmembership via API retains field values for this user = OK
5. saving value of 1 character saves okay
6. saving value 0 saves okay<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19884">issues.joomla.org/tracker/joomla-cms/19884</a>.</sub>

Ruud68 commented:
> Pr's from @Ruud68 got merged.

all credits for my last PR's go to @ggppdk. Some say they where inspired, I just copied his code :)))

Ruud68 commented:
@Quy and @ggppdk can you find some time to test and if okay approve?

Quy commented:
I have tested this item :white_check_mark: successfully on 2756bb8ba6307f20f5769e7ff582379a037330b6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19884">issues.joomla.org/tracker/joomla-cms/19884</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19884">issues.joomla.org/tracker/joomla-cms/19884</a>.</sub>
